### PR TITLE
Python 3 Support

### PR DESCRIPTION
**Pull request by eukreign**

### CODE REVIEW

**Comment by eukreign** [comment]
@jackrobison @shyba this is ready for review.

File: build.sh [hunk 1]
```diff
@@ -1,8 +1,4 @@
 #!/bin/bash
-
 set -euxo pipefail
-
 echo "Building protobuf files"
-rm -rf ./lbryschema/schema/*_pb2.py
```
Comment (jackrobison) [inline build.sh]: I don't think protoc overwrites existing _pb2.py files, this is useful for development

File: lbryschema/proto/claim_pb2.py [hunk 2]
```diff
@@ -13,18 +13,18 @@
...
+import lbryschema.proto.stream_pb2
+import lbryschema.proto.certificate_pb2
+import lbryschema.proto.signature_pb2
 
 
 DESCRIPTOR = _descriptor.FileDescriptor(
-  name='claim.proto',
+  name='lbryschema/proto/claim.proto',
   package='',
-  syntax='proto2',
```
Comment (jackrobison) [inline lbryschema/proto/claim_pb2.py]: Why did this get removed?

Comment (eukreign) [reply lbryschema/proto/claim_pb2.py]: Isn't that generated? I don't recall making that change myself.

**Comment by eukreign** [comment]
@jackrobison please continue with the review but don't merge it yet. i found some issues with it while debugging the publish performance issue.

File: build.sh (hunk 1 above)
Comment (eukreign) [reply build.sh]: It does overwrite.

**Comment by jackrobison** [comment]
@eukreign please test this with lbryum server (building the db using this version of lbryschema), then let's merge it

**Comment by shyba** [comment]
After running against 117k claims db dump 2 failed on both python3 and python2:
```
py3 differ at aa971a70177d97db5422f1864fe9825480526c68: [True, {'version': '_0_0_1', 'claimType': 'streamType', 'stream': {'source': {'source': "\xa0\xa3/|B(>\x0fP\x93\xed\xfe3\xfe/V:\xceaw\x7f\xc9\xf6\x01J\x14\xd55/\xfdN9Kl'\x81\xd9%\xbb\x95\xfd\x1c\xcfGDn\x81u", 'version': '_0_0_1', 'contentType': 'image/png', 'sourceType': 'lbry_sd_hash'}, 'version': '_0_0_1', 'metadata': {'license': 'None', 'description': 'fdsfdsa', 'language': 'en', 'title': 'frfrfrr', 'author': '', 'version': '_0_1_0', 'nsfw': False, 'licenseUrl': '', 'preview': '', 'thumbnail': ''}}}] != [False, 'decode error']
py2 differ at aa971a70177d97db5422f1864fe9825480526c68: [True, {'version': '_0_0_1', 'claimType': 'streamType', 'stream': {'source': {'source': "\xa0\xa3/|B(>\x0fP\x93\xed\xfe3\xfe/V:\xceaw\x7f\xc9\xf6\x01J\x14\xd55/\xfdN9Kl'\x81\xd9%\xbb\x95\xfd\x1c\xcfGDn\x81u", 'version': '_0_0_1', 'contentType': 'image/png', 'sourceType': 'lbry_sd_hash'}, 'version': '_0_0_1', 'metadata': {'license': 'None', 'description': 'fdsfdsa', 'language': 'en', 'title': 'frfrfrr', 'author': '', 'version': '_0_1_0', 'nsfw': False, 'licenseUrl': '', 'preview': '', 'thumbnail': ''}}}] != [False, 'decode error']
py3 differ at 0b1d46fe7e8c1e44d7749fef176b613060fbed14: [True, {'version': '_0_0_1', 'claimType': 'streamType', 'stream': {'source': {'source': '\xe1\x0e\x132\xd1-V\x1f\xe4\xb1?t\xc7\xa7\x8d\xbdy\x07\x14R\xd6\xea=ax&\xfe\xde[\x94\x065\x04R\x92\x1d\x89ytq\x95tF\xb0\xd4\xc8S\x8a', 'version': '_0_0_1', 'contentType': 'image/jpeg', 'sourceType': 'lbry_sd_hash'}, 'version': '_0_0_1', 'metadata': {'license': 'None', 'description': 'dsf', 'language': 'en', 'title': 'test', 'author': '', 'version': '_0_1_0', 'nsfw': False, 'licenseUrl': '', 'preview': '', 'thumbnail': ''}}}] != [False, 'decode error']
py2 differ at 0b1d46fe7e8c1e44d7749fef176b613060fbed14: [True, {'version': '_0_0_1', 'claimType': 'streamType', 'stream': {'source': {'source': '\xe1\x0e\x132\xd1-V\x1f\xe4\xb1?t\xc7\xa7\x8d\xbdy\x07\x14R\xd6\xea=ax&\xfe\xde[\x94\x065\x04R\x92\x1d\x89ytq\x95tF\xb0\xd4\xc8S\x8a', 'version': '_0_0_1', 'contentType': 'image/jpeg', 'sourceType': 'lbry_sd_hash'}, 'version': '_0_0_1', 'metadata': {'license': 'None', 'description': 'dsf', 'language': 'en', 'title': 'test', 'author': '', 'version': '_0_1_0', 'nsfw': False, 'licenseUrl': '', 'preview': '', 'thumbnail': ''}}}] != [False, 'decode error']
```

**Comment by tzarebczan** [comment]
@shyba / @eukreign what's the next step here? Ready to merge after the recent commits?

**Comment by eukreign** [comment]
There are two things delaying this merge:

1. This doesn't bring new functionality to existing daemon or server and actually creates a risk, therefore delaying isn't a bad thing (more time to find bugs?). This pull request is only needed for the new python 3 lbryumx server (electrumx) which is still in early stages of development.

2. I remember having issues with this code when running regular lbrynet daemon and just using the app (clicking on videos, etc). I would like to do some more manual testing and record any failures as unit tests to make sure there is no such future regressions.

After 2 is done and we're closer to releasing electrumx I think the urgency to merge this will be higher.

**Comment by shyba** [comment]
@eukreign @jackrobison This is the quick script I created for checking `smart_decode`. It can easily be extended for checking signature validation or anything else against a lbryum-server dump:
https://gist.github.com/shyba/c28b791230ebc90eac6dd5719910a37f
(should I add it as a PR?)

Parsing is the same everywhere.
We need an RC with it to test on spee.ch and other places.

```bash
(lbry) ➜  lbryum_db python regression.py generate test_smart_decode claim_values tox_and_optimizing
DB size: 112825
Processing item 230 of 112825 0.20385552847/home/ful0n/lbry/lbryschema/lbryschema/claim.py:150: RuntimeWarning: Unexpected end-group tag: Not all data was converted
  temp_claim.ParseFromString(serialized)
Processing item 112824 of 112825 99.9991136716%
Done reading.

Test completed.

$ ... changes to tox branch
(lbry) ➜  lbryum_db python regression.py generate test_smart_decode claim_values master
DB size: 112825
Processing item 230 of 112825 0.20385552847/home/ful0n/lbry/lbryschema/lbryschema/claim.py:149: RuntimeWarning: Unexpected end-group tag: Not all data was converted
  temp_claim.ParseFromString(serialized)
Processing item 112824 of 112825 99.9991136716%
Done reading.

Test completed.
(lbry) ➜  lbryum_db python regression.py generate test_smart_decode claim_values tox               
DB size: 112825
Processing item 230 of 112825 0.20385552847/home/ful0n/lbry/lbryschema/lbryschema/claim.py:150: RuntimeWarning: Unexpected end-group tag: Not all data was converted
  temp_claim.ParseFromString(serialized)
Processing item 112824 of 112825 99.9991136716%
Done reading.

Test completed.

$ .... changes to master branch...
(lbry) ➜  lbryum_db python regression.py compare test_smart_decode
Checking for differences on 3 runs.
Branch tox_and_optimizing has 112825 items.
Branch tox has 112825 items.
Branch master has 112825 items.
Comparing tox_and_optimizing to ['tox', 'master'].
No difference between branches
```

**Comment by tzarebczan** [comment]
@shyba we should add it here or possibly https://github.com/lbryio/jacks-scripts (and rename that repo just to internal scripts?). 

@jackrobison / @eukreign can we get this merged to test on spee.ch?

**Comment by shyba** [comment]
@tzarebczan It's currently deployed to staging spee.ch, there is an RC with it

**Pull request merged**